<a href="https://colab.research.google.com/github/AlirezaZarrinmehr/Intro-to-Data-Science/blob/main/IST%20687%20HW3.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Intro to Data Science HW 3

##### Copyright 2022, Jeffrey Stanton, Jeffrey Saltz, and Jasmina Tacheva

In [70]:
# Enter your name here: 

### Attribution statement: (choose only one and delete the rest)

In [71]:
# 1. I did this homework by myself, with help from the book and the professor.
# 2. I did this homework with help from the book and the professor and these Internet sources:
# 3. I did this homework with help from <Name of another student> but did not cut and paste any code.

### Reminders of things to practice from last week: 
Make a data frame:		data.frame( ) <br>
Row index of max/min:	which.max( )  which.min( )<br>
Sort value or order rows: arrange( )	sort( )   order( )<br>
Descriptive statistics: 	mean( ) sum( ) max( ) <br>
Conditional statement:	if (condition) “true stuff” else “false stuff”<br>

### This Week: 
Often, when you get a dataset, it is not in the format you want. You can (and should) use code to refine the dataset to become more useful. As Chapter 6 of Introduction to Data Science mentions, this is called **“data munging.”** In this homework, you will read in a dataset from the web and work on it (in a data frame) to improve its usefulness.


## Part 1: Use read_csv( ) to read a CSV file from the web into a data frame:

A.	Use R code to read directly from a URL on the web. Store the dataset into a new dataframe, called **testDF**. <br>
The URL is:    <br>
"https://data-science-intro.s3.us-east-2.amazonaws.com/NYS_COVID_Testing.csv" <br>
**Hint:** use read_csv( ), not read.csv( ). This is from the **tidyverse package**. Check the help to compare them.


In [72]:
library(tidyverse)
testDF <- read_csv("https://data-science-intro.s3.us-east-2.amazonaws.com/NYS_COVID_Testing.csv")

Rows: 7383 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): TestDate, AgeGroup, AgeCategory
dbl (2): PositiveCases, TotalTests

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


B.	Use View( ), head( ), and tail( ) to examine the **testDF** dataframe. <br>
**Add a block comment that briefly describes what you see.**

In [73]:
View(testDF)
head(testDF)
tail(testDF)

TestDate AgeGroup PositiveCases TotalTests AgeCategory       
1    3/2/2020 45 to 54 1              1         middle-aged_adults
2    3/3/2020 25 to 34 0              2         young_adults      
3    3/3/2020 35 to 44 0              1         middle-aged_adults
4    3/3/2020 45 to 54 0              1         middle-aged_adults
5    3/3/2020 55 to 64 0              2         senior_citizens   
6    3/3/2020 65 to 74 0              2         senior_citizens   
7    3/3/2020 85 +     0              1         senior_citizens   
8    3/4/2020 1 to 4   0              1         children          
9    3/4/2020 25 to 34 0              5         young_adults      
10   3/4/2020 35 to 44 2              2         middle-aged_adults
11   3/4/2020 45 to 54 0              2         middle-aged_adults
12   3/4/2020 5 to 19  0              3         children          
13   3/4/2020 55 to 64 0              2         senior_citizens   
14   3/4/2020 65 to 74 0              3         senior_citizens   
15   3/4/2020 85 +     0              1         senior_citizens   
16   3/4/2020 < 1      0              1         children          
17   3/5/2020 1 to 4   0              7         children          
18   3/5/2020 20 to 24 1              4         young_adults      
19   3/5/2020 25 to 34 0             10         young_adults      
20   3/5/2020 35 to 44 6             13         middle-aged_adults
21   3/5/2020 45 to 54 4             13         middle-aged_adults
22   3/5/2020 5 to 19  6             17         children          
23   3/5/2020 55 to 64 1             12         senior_citizens   
24   3/5/2020 65 to 74 3              8         senior_citizens   
25   3/5/2020 75 to 84 1              5         senior_citizens   
26   3/5/2020 85 +     0              1         senior_citizens   
27   3/5/2020 < 1      0              2         children          
28   3/6/2020 20 to 24 1              4         young_adults      
29   3/6/2020 25 to 34 1              8         young_adults      
30   3/6/2020 35 to 44 2              9         middle-aged_adults
⋮    ⋮        ⋮        ⋮             ⋮          ⋮                 
7354 1/1/2022 35 to 44 10342         44195      middle-aged_adults
7355 1/1/2022 45 to 54  8796         38519      middle-aged_adults
7356 1/1/2022 5 to 19  10426         39321      children          
7357 1/1/2022 55 to 64  7032         34731      senior_citizens   
7358 1/1/2022 65 to 74  3476         19433      senior_citizens   
7359 1/1/2022 75 to 84  1487          8929      senior_citizens   
7360 1/1/2022 85 +       718          5167      senior_citizens   
7361 1/1/2022 < 1        746          2195      children          
7362 1/2/2022 1 to 4    1775          7818      children          
7363 1/2/2022 20 to 24  4297         17220      young_adults      
7364 1/2/2022 25 to 34  9666         41402      young_adults      
7365 1/2/2022 35 to 44  8796         35585      middle-aged_adults
7366 1/2/2022 45 to 54  7158         30384      middle-aged_adults
7367 1/2/2022 5 to 19   8965         36061      children          
7368 1/2/2022 55 to 64  5868         27295      senior_citizens   
7369 1/2/2022 65 to 74  2755         14931      senior_citizens   
7370 1/2/2022 75 to 84  1195          6808      senior_citizens   
7371 1/2/2022 85 +       522          3754      senior_citizens   
7372 1/2/2022 < 1        700          1894      children          
7373 1/3/2022 1 to 4    1857          8794      children          
7374 1/3/2022 20 to 24  4380         18944      young_adults      
7375 1/3/2022 25 to 34  9875         46786      young_adults      
7376 1/3/2022 35 to 44  8840         37432      middle-aged_adults
7377 1/3/2022 45 to 54  7361         31822      middle-aged_adults
7378 1/3/2022 5 to 19   9923         38977      children          
7379 1/3/2022 55 to 64  5739         27019      senior_citizens   
7380 1/3/2022 65 to 74  2759         14498      senior_citizens   
7381 1/3/2022 75 to 84  1141          6519      seni

TestDate AgeGroup PositiveCases TotalTests AgeCategory       
1 3/2/2020 45 to 54 1             1          middle-aged_adults
2 3/3/2020 25 to 34 0             2          young_adults      
3 3/3/2020 35 to 44 0             1          middle-aged_adults
4 3/3/2020 45 to 54 0             1          middle-aged_adults
5 3/3/2020 55 to 64 0             2          senior_citizens   
6 3/3/2020 65 to 74 0             2          senior_citizens

TestDate AgeGroup PositiveCases TotalTests AgeCategory    
1 1/3/2022 5 to 19  9923          38977      children       
2 1/3/2022 55 to 64 5739          27019      senior_citizens
3 1/3/2022 65 to 74 2759          14498      senior_citizens
4 1/3/2022 75 to 84 1141           6519      senior_citizens
5 1/3/2022 85 +      680           4028      senior_citizens
6 1/3/2022 < 1       717           2074      children

## Part 2: Create new data frames based on a condition:

A.	Use the table( ) command to summarize the contents of the **AgeCategory** variable in **testDF**. <br>
**Write a comment interpreting what you see – how many age categories are there in the dataset and what is the proportion of observations in each?**

In [74]:
table(testDF$AgeCategory)


          children middle-aged_adults    senior_citizens       young_adults 
              2010               1345               2686               1342 

B. Terms like "senior citizens" can function as *othering* language which demeans the people it seeks to describe. 

We can use the **str_replace_all()** function from tidyverse to find all instances of **senior_citizens** in the **AgeCategory** variable and replace them with **older_adults**. 

In this case, we want to search for **senior_citizens** and replace it with **older_adults** in **testDF$AgeCategory** - how can you use this information to overwrite the **AgeCategory** in the function below:

In [75]:
library(tidyverse)
testDF$AgeCategory <- str_replace_all(testDF$AgeCategory, "senior_citizens", "older_adults")
head(testDF)

TestDate AgeGroup PositiveCases TotalTests AgeCategory       
1 3/2/2020 45 to 54 1             1          middle-aged_adults
2 3/3/2020 25 to 34 0             2          young_adults      
3 3/3/2020 35 to 44 0             1          middle-aged_adults
4 3/3/2020 45 to 54 0             1          middle-aged_adults
5 3/3/2020 55 to 64 0             2          older_adults      
6 3/3/2020 65 to 74 0             2          older_adults

C.	Create a dataframe (called **olderAdults**) that contains only the rows (observations) for which the value in the **AgeCategory** variable (column) is **older_adults**. Hint: Use subsetting.

In [76]:
library(dplyr)
olderAdults <- subset(testDF, AgeCategory == "older_adults")
head(olderAdults)

TestDate AgeGroup PositiveCases TotalTests AgeCategory 
1 3/3/2020 55 to 64 0             2          older_adults
2 3/3/2020 65 to 74 0             2          older_adults
3 3/3/2020 85 +     0             1          older_adults
4 3/4/2020 55 to 64 0             2          older_adults
5 3/4/2020 65 to 74 0             3          older_adults
6 3/4/2020 85 +     0             1          older_adults

D.	Use the dim() command on **olderAdults** to confirm that the data frame contains **2,686** observations and **5** columns/variables.

In [77]:
dim(olderAdults)

[1] 2686    5

E.	Use **subsetting** to create a new dataframe that contains only the observations for which the value in the **AgeCategory** variable is **young_adults**. The name of this new df should be **youngAdults**.

In [78]:
youngAdults <- subset(testDF, AgeCategory == "young_adults")
head(youngAdults)

TestDate AgeGroup PositiveCases TotalTests AgeCategory 
1 3/3/2020 25 to 34 0              2         young_adults
2 3/4/2020 25 to 34 0              5         young_adults
3 3/5/2020 20 to 24 1              4         young_adults
4 3/5/2020 25 to 34 0             10         young_adults
5 3/6/2020 20 to 24 1              4         young_adults
6 3/6/2020 25 to 34 1              8         young_adults

F. Create one last data frame which only contains the observations for **children** in the **AgeCategory** variable of **testDF**. Call this new df **childrenDF**.

In [79]:
childrenDF <- subset(testDF, AgeCategory == "children")
head(childrenDF)

TestDate AgeGroup PositiveCases TotalTests AgeCategory
1 3/4/2020 1 to 4   0              1         children   
2 3/4/2020 5 to 19  0              3         children   
3 3/4/2020 < 1      0              1         children   
4 3/5/2020 1 to 4   0              7         children   
5 3/5/2020 5 to 19  6             17         children   
6 3/5/2020 < 1      0              2         children

## Part 3: Analyze the numeric variables in the **testDF** dataframe.

A.	How many **numeric variables** does the dataframe have? You can figure that out by looking at the output of **str(testDF)**. 

In [80]:
str(testDF)

spec_tbl_df [7,383 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ TestDate     : chr [1:7383] "3/2/2020" "3/3/2020" "3/3/2020" "3/3/2020" ...
 $ AgeGroup     : chr [1:7383] "45 to 54" "25 to 34" "35 to 44" "45 to 54" ...
 $ PositiveCases: num [1:7383] 1 0 0 0 0 0 0 0 0 2 ...
 $ TotalTests   : num [1:7383] 1 2 1 1 2 2 1 1 5 2 ...
 $ AgeCategory  : chr [1:7383] "middle-aged_adults" "young_adults" "middle-aged_adults" "middle-aged_adults" ...
 - attr(*, "spec")=
  .. cols(
  ..   TestDate = col_character(),
  ..   AgeGroup = col_character(),
  ..   PositiveCases = col_double(),
  ..   TotalTests = col_double(),
  ..   AgeCategory = col_character()
  .. )
 - attr(*, "problems")=<externalptr> 


B.	What is the average number of total daily tests? Hint: Can you think of a mathematical function we've come across before to use on the **TotalTests** variable?

In [81]:
mean(testDF$TotalTests)

[1] 11724.96

C.	How many tests were performed in the row with the highest number of total daily tests? What age category do they correspond to? 

In [82]:
max(testDF$TotalTests)
testDF[testDF$TotalTests == max(testDF$TotalTests),]

[1] 88904

TestDate   AgeGroup PositiveCases TotalTests AgeCategory 
1 12/26/0021 25 to 34 12588         88904      young_adults

D. How many positive cases were registered in the row with the highest number of positive cases? What age category do they correspond to?

In [83]:
max(testDF$PositiveCases)
testDF[testDF$PositiveCases == max(testDF$PositiveCases),]

[1] 17486

TestDate   AgeGroup PositiveCases TotalTests AgeCategory 
1 12/31/2021 25 to 34 17486         78574      young_adults

E.	What is the total number of positive cases in **testDF**?

In [84]:
sum(testDF$PositiveCases)

[1] 3722542

F. Create a new variable in **testDF** which is the ratio of **PostiveCases** to **TotalTests**. Call this variable **PositivityRate** and explain in a comment what information it gives us.

In [86]:
testDF$PositivityRate<-testDF$PositiveCases/testDF$TotalTests


G. What is the average positivity rate in **testDF**? Hint: Use the **mean()** function on the new variable you created in F.

In [88]:
mean(testDF$PositivityRate)

[1] 0.05363633

## Part 4:  Create a function to automate the process from F-G:

A.	The following function should work most of the time. Make sure to run this code before trying to test it. That is how you make the new function known to R. **Add comments to each line explaining what it does:**

In [89]:
calculatePositivity <- function(dataset) {
  dataset$PositivityRate <- dataset$PositiveCases/dataset$TotalTests
  avePositivity <- mean(dataset$PositivityRate)
  return(avePositivity)
}

B.	Run your new function on the **testDF** dataframe. Is the output of the function consistent with the output in Step G above? Explain.

In [90]:
calculatePositivity(testDF)

[1] 0.05363633

C. Run the function on the **olderAdults** df you created earlier.

In [91]:
calculatePositivity(olderAdults)

[1] 0.05386202

D.	Run the function on the **youngAdults** df.

In [92]:
calculatePositivity(youngAdults)

[1] 0.05435627

E. Lastly, run the posivity function on the **childrenDF** dataframe.

In [93]:
calculatePositivity(childrenDF)

[1] 0.05034919

F. In a comment, describe what you observe across these 3 datasets - which age group exhibits the highest positivity rate? How do these numbers compare to the **baseline** positivity rate in **testDF**?